src/app/ClusterInfo.h

```diff
@@ -39,15 +42,44 @@ struct ClusterInfo
             return false;
         }
 
-        if (mFlags.Has(Flags::kFieldIdValid) && (mFieldId == 0xFFFFFFFF))
+        Optional<AttributeId> myFieldId =
+            mFlags.Has(Flags::kFieldIdValid) ? Optional<AttributeId>::Value(mFieldId) : Optional<AttributeId>::Missing();
+
+        Optional<AttributeId> otherFieldId = other.mFlags.Has(Flags::kFieldIdValid) ? Optional<AttributeId>::Value(other.mFieldId)
+                                                                                    : Optional<AttributeId>::Missing();
+
+        Optional<ListIndex> myListIndex =
+            mFlags.Has(Flags::kListIndexValid) ? Optional<ListIndex>::Value(mListIndex) : Optional<ListIndex>::Missing();
+
+        Optional<ListIndex> otherListIndex = other.mFlags.Has(Flags::kListIndexValid) ? Optional<ListIndex>::Value(other.mListIndex)
+                                                                                      : Optional<ListIndex>::Missing();
+
+        if (myFieldId == Optional<AttributeId>::Value(kRootAttributeId))
+        {
+            return true;
+        }
+
+        if (myFieldId != otherFieldId)
+        {
+            return false;
+        }
+
+        // We only support top layer for attribute representation, either FieldId or FieldId + ListIndex
+        // Combination: if myFieldId == otherFieldId, ListIndex cannot exist without FieldId
+        // 1. myListIndex and otherListIndex both missing or both exactly the same, then current is superset of other
+        // 2. myListIndex is missing, no matter if otherListIndex is missing or not, then current is superset of other
+        if (myListIndex == otherListIndex)
         {
+            // either both missing or both exactly the same
             return true;
         }
 
-        if (mFlags.Has(Flags::kFieldIdValid) && other.mFlags.Has(Flags::kFieldIdValid) && (mFieldId == other.mFieldId))
+        if (!myListIndex.HasValue())
```

Comment:
`return !myListIndex.HasValue()` seems to be the last line.

However then why does it return false if myListIndexHaveValue?
I expect 2 identical values (with list index) to return true.

Please add unit tests for the possible combinations of things here.

Reply:
I think the code is  correct: if we get here, the list indices are different. May be worht a comment: difference ok only if mylistindex is missing (missing is a superset of specific)

Reply:
update comment, and the test has been added for these possible combination, thanks